File: leetcode/Easy/20. Valid Parentheses.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        
        for (char ch : s) {
            if (ch == '(' || ch == '{' || ch == '[') {
                st.push(ch);
            } else {
                if (st.empty()) {
                    return false;
                } else {
                    if (ch == ')') {
                        if (st.top() != '(') {
                            return false;
                        } else if (st.top() == '(') {
                            st.pop();
                        }
                    } else if (ch == '}') {
                        if (st.top() != '{') {
                            return false;
                        } else if (st.top() == '{') {
                            st.pop();
                        }
                    } else if (ch == ']') {
                        if (st.top() != '[') {
                            return false;
                        } else if (st.top() == '[') {
                            st.pop();
                        }
                    }
                }
            }
        }
        if (!st.empty()) {
            return false;
        } else {
            return true;
        }
    }
};

File: leetcode/Easy/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        dic = {')':'(', '}':'{', ']':'['}
        
        for ch in s:
            if ch in dic.values():
                st.append(ch)
            elif ch in dic.keys():
                if st == [] or dic[ch] != st.pop():
                    return False
            else:
                return False
        return st == []